iijpta/ngx_http_secure_token_iijpta.c

```diff
@@ -0,0 +1,231 @@
+#include <ngx_config.h>
+
+#include "ngx_http_secure_token_iijpta.h"
+#include "../ngx_http_secure_token_filter_module.h"
+#include "../ngx_http_secure_token_utils.h"
+
+#include <openssl/evp.h>
+
+// constants
+#define CRC32_SIZE    4
+#define EXPIRY_SIZE   8
+#define PATH_LIMIT    1024
+
+static char *ngx_conf_check_byte_len_bounds(ngx_conf_t *cf, void *post, void *data);
+
+// typedefs
+typedef struct {
+	ngx_str_t key;
+	ngx_str_t iv;
+	ngx_str_t path;
+	ngx_secure_token_time_t end;
+} ngx_secure_token_iijpta_token_t;
+
+typedef struct {
+        u_char crc[CRC32_SIZE];
+        u_char expiry[EXPIRY_SIZE];
+} ngx_http_secure_token_iijpta_header_t;
+
+static ngx_conf_num_bounds_t ngx_http_secure_token_iijpta_key_bounds = {
+	ngx_conf_check_byte_len_bounds, 16, 16
+};
+
+static ngx_conf_num_bounds_t ngx_http_secure_token_iijpta_iv_bounds = {
+	ngx_conf_check_byte_len_bounds, 16, 16
+};
+
+// globals
+static ngx_command_t ngx_http_secure_token_iijpta_cmds[] = {
+	{ ngx_string("key"),
+	NGX_CONF_TAKE1,
+	ngx_http_secure_token_conf_set_hex_str_slot,
+	0,
+	offsetof(ngx_secure_token_iijpta_token_t, key),
+	&ngx_http_secure_token_iijpta_key_bounds },
+
+	{ ngx_string("iv"),
+	NGX_CONF_TAKE1,
+	ngx_http_secure_token_conf_set_hex_str_slot,
+	0,
+	offsetof(ngx_secure_token_iijpta_token_t, iv),
+	&ngx_http_secure_token_iijpta_iv_bounds },
+
+	{ ngx_string("path"),
+	NGX_CONF_TAKE1,
+	ngx_conf_set_str_slot,
+	0,
+	offsetof(ngx_secure_token_iijpta_token_t, path),
+	NULL },
+
+	{ ngx_string("end"),
+	NGX_CONF_TAKE1,
+        ngx_http_secure_token_conf_set_time_slot,
+	0,
+	offsetof(ngx_secure_token_iijpta_token_t, end),
+	NULL },
+};
+
+static char *
+ngx_conf_check_byte_len_bounds(ngx_conf_t *cf, void *post, void *data)
+{
+	ngx_conf_num_bounds_t  *bounds = post;
+	ngx_str_t  *sp = data;
+
+	if (bounds->high == -1) {
+		if (sp->len >= (size_t)bounds->low) {
+			return NGX_CONF_OK;
+		}
+
+		ngx_conf_log_error(NGX_LOG_EMERG, cf, 0,
+			"value must be equal to or greater than %i",
+			bounds->low);
+
+		return NGX_CONF_ERROR;
+	}
+
+	if (sp->len >= (size_t)bounds->low && sp->len <= (size_t)bounds->high) {
+		return NGX_CONF_OK;
+	}
+
+	ngx_conf_log_error(NGX_LOG_EMERG, cf, 0,
+		"value must be between %i byte and %i byte",
+		bounds->low, bounds->high);
+
+	return NGX_CONF_ERROR;
+}
+
+static ngx_int_t
+ngx_secure_token_iijpta_get_var(
+	ngx_http_request_t *r,
+	ngx_http_variable_value_t *v,
+	uintptr_t data)
+{
+	EVP_CIPHER_CTX *ctx;
+	uint32_t crc;
+	ngx_secure_token_iijpta_token_t* token = (void*)data;
+	u_char* in;
+	ngx_http_secure_token_iijpta_header_t *hdr;
+	size_t in_len  = sizeof(ngx_http_secure_token_iijpta_header_t) + token->path.len;
+	u_char *p;
+	// in_len rounded up to block + one block for padding
+	uint8_t out[((in_len + (16 - 1)) / 16) + 16];
```

Comment:
Some compilers may not accept this (dynamically sized array) better allocate with ngx_pnalloc